kernel/src/grant.rs

```diff
@@ -56,155 +27,170 @@ impl<T: ?Sized> Owned<T> {
-        }
-    }
-
-    /// Allocates a slice of n instances of a given type. Each instance is
-    /// initialized using the provided function.
-    ///
-    /// The provided function will be called exactly `n` times, and will be
-    /// passed the index it's initializing, from `0` through `num_items - 1`.
-    ///
-    /// # Panic Safety
-    ///
-    /// If `val_func` panics, the freshly allocated memory and any values
-    /// already written will be leaked.
-    pub fn alloc_n_with<T, F>(
-        &mut self,
-        num_items: usize,
-        mut val_func: F,
-    ) -> Result<Owned<[T]>, Error>
-    where
-        F: FnMut(usize) -> T,
-    {
-        unsafe {
-            let ptr = self.alloc_n_raw::<T>(num_items)?;
+pub struct AppliedGrant<'a, T: 'a> {
+    process: &'a dyn ProcessType,
+    grant_num: usize,
+    grant: &'a mut T,
+    _phantom: PhantomData<T>,
+}
 
-            for i in 0..num_items {
-                write(ptr.as_ptr().add(i), val_func(i));
+impl<'a, T: Default> AppliedGrant<'a, T> {
+    fn get_or_allocate(grant: &Grant<T>, process: &'a dyn ProcessType) -> Result<Self, Error> {
+        // Here is an example of how the grants are laid out in a
+        // process's memory:
+        //
+        // Mem. Addr.
+        // 0x0040000  ┌────────────────────
+        //            │   GrantPointer0 [0x003FFC8]
+        //            │   GrantPointer1 [0x003FFC0]
+        //            │   ...
+        //            │   GrantPointerN [0x0000000 (NULL)]
+        //            ├────────────────────
+        //            │   Process Control Block
+        // 0x003FFE0  ├────────────────────
+        //            │   GrantRegion0
+        // 0x003FFC8  ├────────────────────
+        //            │   GrantRegion1
+        // 0x003FFC0  ├────────────────────
+        //            │
+        //            │   --unallocated--
+        //            │
+        //            └────────────────────
+        //
+        // An array of pointers (one per possible grant region)
+        // point to where the actual grant memory is allocated
+        // inside of the process. The grant memory is not allocated
+        // until the actual grant region is actually used.
+        //
+        // This function provides the app access to the specific
+        // grant memory, and allocates the grant region in the
+        // process memory if needed.
+
+        // Get the GrantPointer to start. Since process.rs does not know
+        // anything about the datatype of the grant, and the grant
+        // memory may not yet be allocated, it can only return a `*mut
+        // u8` here. We will eventually convert this to a `*mut T`.
+        let grant_num = grant.grant_num;
+        let appid = process.appid();
+        if let Some(untyped_grant_ptr) = process.get_grant_ptr(grant_num) {
+            // If the grant pointer is NULL then the memory for the
+            // GrantRegion needs to be allocated. Otherwise, we can
```

Comment:
```suggestion
            // GrantRegion needs to be allocated. If the grant pointer is all 1s
            // then the grant is already enetered and we return an error. Otherwise, we can
```